### PR TITLE
Add XPathExtractionOutputFilter

### PR DESCRIPTION
**Pull request by gcdev373**
Some problems were identified when importing the test data.

### CODE REVIEW

**Comment by gcdev373** [comment]
Fixes issue #818

**Comment by gcdev373** [comment]
Not ready! Needs modification.

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionElementModule.java [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package stroom.search.extraction;
+
+import stroom.pipeline.factory.PipelineElementModule;
+
+class XPathExtractionElementModule extends PipelineElementModule {
```
Comment (at055612) [inline stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionElementModule.java]: Not sure it needs its own module, just add the bind to the ExtractionElementModule

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 2]
```diff
@@ -0,0 +1,420 @@
...
+    private Locator locator;
+
+    private final Configuration config = new Configuration();
+    private final PipelineConfiguration pipeConfig = config.makePipelineConfiguration();
+    final Processor processor = new Processor(config);
+    final XPathCompiler compiler = processor.newXPathCompiler();
+
+    private TinyBuilder builder = null;
+
+    private ReceivingContentHandler rch = null;
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: OK

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 3]
```diff
@@ -0,0 +1,420 @@
...
+import javax.inject.Inject;
+import java.util.HashMap;
+import java.util.Iterator;
+
+import static stroom.index.shared.IndexConstants.EVENT_ID;
+import static stroom.index.shared.IndexConstants.STREAM_ID;
+
+
+@ConfigurableElement(type = "XPathExtractionOutputFilter", category = Category.FILTER, roles = {
+        PipelineElementType.ROLE_TARGET, PipelineElementType.ROLE_HAS_TARGETS}, icon = ElementIcons.SEARCH)
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: ROLE_HAS_TARGETS removed.

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 4]
```diff
@@ -0,0 +1,420 @@
...
+
+        super.endElement(uri, localName, qName);
+    }
+
+    @Override
+    public void characters(char[] ch, int start, int length) throws SAXException {
+        if (rch != null)
+            rch.characters(ch, start, length);
+        else
+            log(Severity.ERROR, "Unexpected text node " + new String(ch, start, length) +
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: OK

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 5]
```diff
@@ -0,0 +1,420 @@
...
+    private final ErrorReceiverProxy errorReceiverProxy;
+    private final SecurityContext securityContext;
+    private final LocationFactoryProxy locationFactory;
+
+    private Locator locator;
+
+    private final Configuration config = new Configuration();
+    private final PipelineConfiguration pipeConfig = config.makePipelineConfiguration();
+    final Processor processor = new Processor(config);
+    final XPathCompiler compiler = processor.newXPathCompiler();
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: Now private.

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 6]
```diff
@@ -0,0 +1,420 @@
...
+    @Override
+    public void startDocument() throws SAXException {
+        depth = 0;
+        super.startDocument();
+    }
+
+
+    private String topLevelUri = null;
+    private String topLevelQName = null;
+    private Attributes topLevelAtts = null;
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: OK

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 7]
```diff
@@ -0,0 +1,420 @@
...
+
+    private TinyBuilder builder = null;
+
+    private ReceivingContentHandler rch = null;
+    private String topLevelElementToSkip = "";
+    private String secondLevelElementToCreateDocs = "";
+    private int depth = 0;
+
+    private HashMap<String, String> prefixMappings = new HashMap<>();
+    private XPathExecutable[] xPathExecutables = null;;
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: OK

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java [hunk 8]
```diff
@@ -0,0 +1,420 @@
...
+
+import static stroom.index.shared.IndexConstants.EVENT_ID;
+import static stroom.index.shared.IndexConstants.STREAM_ID;
+
+
+@ConfigurableElement(type = "XPathExtractionOutputFilter", category = Category.FILTER, roles = {
+        PipelineElementType.ROLE_TARGET, PipelineElementType.ROLE_HAS_TARGETS}, icon = ElementIcons.SEARCH)
+public class XPathExtractionOutputFilter extends SearchResultOutputFilter {
+    private String multipleValueDelimiter = DEFAULT_MULTIPLE_STRING_DELIMITER;
+    private static final String DEFAULT_MULTIPLE_STRING_DELIMITER = ",";
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: OK

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionElementModule.java (hunk 1 above)
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionElementModule.java]: OK

File: stroom-db-util/src/main/java/stroom/db/util/DbUtil.java [hunk 9]
```diff
@@ -19,7 +19,7 @@
 
 public class DbUtil {
     private static final Logger LOGGER = LoggerFactory.getLogger(DbUtil.class);
-    private static final boolean USE_TEST_CONTAINERS = true;
+    private static final boolean USE_TEST_CONTAINERS = false;
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-db-util/src/main/java/stroom/db/util/DbUtil.java]: OK

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/XPathExtractionOutputFilter.java]: Need icon added

File: stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/SearchResultOutputFilter.java [hunk 10]
```diff
@@ -36,37 +36,46 @@
     private static final String NAME = "name";
     private static final String VALUE = "value";
 
-    private FieldIndexMap fieldIndexes;
-    private ResultReceiver resultReceiver;
-    private Val[] values;
+    protected FieldIndexMap fieldIndexes;
+    protected ResultReceiver resultReceiver;
+    protected Val[] values;
```
[the remark this answers is not in the document]
Comment (gcdev373) [reply stroom-search/stroom-search-extraction/src/main/java/stroom/search/extraction/SearchResultOutputFilter.java]: Created an abstract base class that both SearchResultOutputFilter and XPathOutputFilter now extend.

**Pull request merged**